managed/services/preconditions.go

```diff
@@ -87,3 +89,85 @@ func CheckMongoDBBackupPreconditions(q *reform.Querier, mode models.BackupMode,
 
 	return nil
 }
+
+// CheckArtifactOverlapping checks if there are other artifacts or scheduled tasks pointing to the same location and folder.
+// Placing MySQL and MongoDB artifacts in the same folder is not desirable, while placing MongoDB artifacts of different clusters
+// in the same folder may cause data inconsistency.
+//
+// WARNING: This function valid only when executed as part of transaction with serializable isolation level.
+func CheckArtifactOverlapping(q *reform.Querier, serviceID, locationID, folder string) error {
+	// TODO This doesn't work for all cases. For example, there are may exist more than one storage locations pointing to the same place.
+
+	service, err := models.FindServiceByID(q, serviceID)
+	if err != nil {
+		return err
+	}
+
+	artifacts, err := models.FindArtifacts(q, models.ArtifactFilters{
+		LocationID: locationID,
+		Folder:     &folder,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, artifact := range artifacts {
+		if artifact.ServiceID != serviceID {
+			svc, err := models.FindServiceByID(q, artifact.ServiceID)
+			if err != nil {
+				return err
+			}
+
+			if service.ServiceType == models.MySQLServiceType && svc.ServiceType == models.MySQLServiceType {
+				continue
+			}
+
+			if service.ServiceType == models.MongoDBServiceType && svc.ServiceType == models.MongoDBServiceType {
+				if svc.Cluster != service.Cluster {
+					return errors.Wrapf(ErrLocationFolderPairAlreadyUsed,
+						"Same location and folder already used for artifact %s of other service: %s", artifact.ID, serviceID)
+				}
+				continue
+			}
+
+			return errors.Wrapf(ErrLocationFolderPairAlreadyUsed,
+				"Same location and folder already used for artifact %s of other service: %s", artifact.ID, serviceID)
```

Comment:
Seems to be the same as line 128. Maybe need a const?

Reply:
Done.